Logic to validate resources during deployment

mskdeepak-oracle opened the pull request:
Fixes #21858 
- `ResourceValidator.java` handles all the validation logic. It gets called upon the event `AFTER_APPLICATION_CLASSLOADER_CREATION`. We need to do this validation after `DEPLOYMENT_BEFORE_CLASSLOADER_CREATION` because application scoped resources are processed at this point
- Used the `DolProvider` class to initialize the `ResourceValidator` as a bit of a hack. Please suggest a better way if possible
- Handled `AppScopedResources` by storing all the relevant JNDI names in the transient app meta data of Depl. Context
                   - Was unable to use the already existing data related to app scoped resources in transient app meta data because of some dependency issues 
- Added devtests corresponding to the handling of resources in `ResourceReferenceDescriptor`, `EnvironmentProperties`, `ResourceEnvReferenceDescriptor`, `MessageDestRefDesc` and for `AppScopedResources`
- Strategy for validating a resource's jndi name
     1) Check in domain.xml
     2) Check for the resources defined within the application. Since these have not been binded to the namespace yet, we cannot do a simple context lookup
               -  I have used the class `JNDINamespace` for this purpose. 
               -  The call to `parseResources` implements this funcitonality 
     3) Check for resources defined by an earlier application by doing a context lookup.

mskdeepak-oracle committed:
Brief outline of resource validation code. Made a hack to get something working. Logic to lookup in config beans added.

mskdeepak-oracle committed:
Extract resources out of dc

mskdeepak-oracle committed:
Fixing some bugs, add validation for references inside an ejb

mskdeepak-oracle committed:
Add validation logic for MessageDestRefDesc

mskdeepak-oracle committed:
Main logic for validating app-scoped resources complete

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_all_resources

mskdeepak-oracle committed:
Rename files because of #21828

mskdeepak-oracle committed:
Wrote devtests for ResourceRef handling code. Handles URL resources

mskdeepak-oracle committed:
Minor comments

mskdeepak-oracle commented:
@glassfishrobot Run CI tests please

mskdeepak-oracle committed:
App scoped resources done

mskdeepak-oracle committed:
Remove target from create custom resource, fix devtest

mskdeepak-oracle commented:
@glassfishrobot Run CI tests please

mskdeepak-oracle commented:
@glassfishrobot Run CI tests please

mskdeepak-oracle commented:
@glassfishrobot Run CI tests please

mskdeepak-oracle committed:
Pass JndiNameEnv instead of module name

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_all_resources

mskdeepak-oracle committed:
Add copyright header

honghzzhang commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,281 @@
+/*
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ *
+ * Copyright (c) 1997-2017 Oracle and/or its affiliates. All rights reserved.
```
For new file, I think the copyright only needs to have the current year (2017). For existing files, you update the range to use the current year.

honghzzhang commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidatorVisitor.java:
```diff
@@ -0,0 +1,51 @@
+/*
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+ *
+ * Copyright (c) 1997-2017 Oracle and/or its affiliates. All rights reserved.
```
Same comment for copyright

honghzzhang reviewed (dismissed):
Thanks for incorporating all my suggestions. I am not sure if there are better alternatives than what you did here without digging into details, but what you had looks fine to me.

mskdeepak-oracle committed:
Fix year in copyright

mskdeepak-oracle commented:
@glassfishrobot Run CI tests please

mskdeepak-oracle committed:
Fix a bug in MessageDest validation

mskdeepak-oracle committed:
Parse all resources and store their names before validating. devtest for message destination resources

mskdeepak-oracle committed:
Devtests for environment properties validation

mskdeepak-oracle committed:
Convert comp namespace to module if needed

mskdeepak-oracle committed:
Null pointer check

mskdeepak-oracle committed:
Add comments

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_all_resources

mskdeepak-oracle committed:
Fix NPE, store resource descriptors in name space

mskdeepak-oracle committed:
Use existing func for parsing resource

mskdeepak-oracle committed:
Validation for PU, MDB

mskdeepak-oracle committed:
Switching off MDB

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_all_resources

mskdeepak-oracle committed:
Managed Beans done. Testing

mskdeepak-oracle committed:
Default names in resource env ref

mskdeepak-oracle committed:
NPE fixes

mskdeepak-oracle committed:
Code cleanup

mskdeepak-oracle committed:
Devtests for Managed Beans, generic resources. ResourceEnvRef done

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_all_resources

mskdeepak-oracle committed:
Move all default checks to one place

mskdeepak-oracle committed:
Find Bugs error

mskdeepak-oracle committed:
Add more types of resources into defaultresources test

mskdeepak-oracle committed:
Merge fixes for #21954, #21955

mskdeepak-oracle committed:
Do a context lookup for resources already existing in global env

mskdeepak-oracle committed:
Fail if null JNDI name

mskdeepak-oracle committed:
Do context lookup only on current instance

mskdeepak-oracle committed:
First commit related to EJB annotation. WIP

mskdeepak-oracle committed:
Fix small bug

mskdeepak-oracle committed:
Code cleanup, fix an NPE

Conflicts:
	appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java

mskdeepak-oracle committed:
Code cleanup, fix an NPE

mskdeepak-oracle committed:
Remove cluster changes

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_all_resources

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_other_annotations

mskdeepak-oracle committed:
Add default resources

mskdeepak-oracle committed:
Add default resources

mskdeepak-oracle committed:
Validate only JMS MDB's

mskdeepak-oracle committed:
Add earlier exception in initcause to show complete stack trace in server.log

mskdeepak-oracle committed:
Comment out ejb ref processing for now

mskdeepak-oracle committed:
Don't validate MQ destinations

mskdeepak-oracle committed:
Cleanup devtests

mskdeepak-oracle committed:
Cleanup devtests

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_other_annotations

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_all_resources

vinayvishal commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+            message = "JNDI lookup failed for the resource with jndi name: {0}",
+            level = "SEVERE",
+            cause = "JNDI lookup for the specified resource failed.",
+            action = "Create the necessary object before deploying the application.",
```
Instead of saying "Create the necessary object before deploying the application.",  I guess stating "Configure the required resources before deploying the application."

vinayvishal commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+    )
+    private static final String RESOURCE_REF_JNDI_LOOKUP_FAILED = "AS-DEPLOYMENT-00026";
+
+    String target;
```
Please make these variables private.

vinayvishal commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+     * Parse all the resources and store them in a namespace before starting the validation.
+     */
+    private void parseResources() {
+        myNamespace = new JNDINamespace();
```
We create a new JNDINamespace() each time this method is getting invoked. Instead of instance variable, local variable can be used.

vinayvishal commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+        myNamespace = new JNDINamespace();
+        parseResources(application);
+        for (BundleDescriptor bd : application.getBundleDescriptorsOfType(DOLUtils.warType())) {
+            parseResources((JndiNameEnvironment) bd);
```
Will `BundleDescriptor` always be an instance of `JndiNameEnvironment`, if not, consider using type check.

vinayvishal commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+        for (BundleDescriptor bd : application.getBundleDescriptorsOfType(DOLUtils.warType())) {
+            parseResources((JndiNameEnvironment) bd);
+        }
+        for (BundleDescriptor bd : application.getBundleDescriptorsOfType(DOLUtils.carType())) {
```
// consider adding comment at places where code itself is not self explanatory.

vinayvishal commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+        for (BundleDescriptor bd : application.getBundleDescriptorsOfType(DOLUtils.carType())) {
+            parseResources((JndiNameEnvironment) bd);
+        }
+        for (BundleDescriptor bd : application.getBundleDescriptorsOfType(DOLUtils.ejbType())) {
```
Instead of having three different loops based on archive type, would it be better to run it once and get the type from `BundleDescriptor` and call relevant method based on the type?

vinayvishal commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+
+    private void storeResource(String name, JndiNameEnvironment env) {
+        String logicalJNDIName = getLogicalJNDIName(name, env);
+        myNamespace.store(logicalJNDIName, env);
```
I see that `myNamespace` is being used in various methods. If we are creating a `ResourceValidator` instance each time an application is being deployed, I guess this should be fine, else if we create only one such instance at server start, we would be better off passing it as a method argument.

mskdeepak-oracle replied:
Yes, changing the code as suggested.

mskdeepak-oracle committed:
Fix some review comments

mskdeepak-oracle committed:
Fix bug

vinayvishal reviewed (dismissed):
ResourceValidator class ideally should be broken down to many smaller sub-classes. If time permits, then now, else we should look at refactoring it at a later stage.

I feel there is an scope of improvement in terms of exception handling, loops, null checks and overall breaking down this class into many smaller subclasses.

Changes are looking good otherwise.

mskdeepak-oracle committed:
Incorporate comments of Nigel and Amy

mskdeepak-oracle commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+    }
+
+    private void parseResources(JndiNameEnvironment env) {
+        for (Object next : env.getResourceReferenceDescriptors()) {
```
No, this is because we are using an upcasted class(JndiNameEnvironment).

mskdeepak-oracle commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+
+    private void parseResources(JndiNameEnvironment env) {
+        for (Object next : env.getResourceReferenceDescriptors()) {
+            parseResources((ResourceReferenceDescriptor) next, env);
```
These are probably unnecessary since the three Bundle Descriptors we use(Web, EJB, App Client) generally return the correct class.

mskdeepak-oracle commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+        }
+
+        /**
+         * TODO: App client doesn't support some resource descriptors. Might need to fail deployment in such cases.
```
Yes. I found some comments in the processing of resource descriptors which suggested the above. Let's discuss about this

mskdeepak-oracle commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+        }
+    }
+
+    private void parseResources(ResourceReferenceDescriptor resRef, JndiNameEnvironment env) {
```
No, this is not possible. I have written separate functions only for `ResourceReferenceDescriptor` and `ResourceEnvReferenceDescriptor` because I wanted to do a `ckeckType` function call for them.
This is not present for the rest of the descriptors. Further, there does not exist such as common interface which can do the functionality you mentioned above.

mskdeepak-oracle commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+    }
+
+    private void accept(EjbDescriptor ejb) {
+        // TODO: MDB
```
Yes.

mskdeepak-oracle commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+            return;
+
+        if (resRef.isURLResource()) {
+            if (jndiName != null && !(jndiName.startsWith(ResourceConstants.JAVA_SCOPE_PREFIX))) {
```
It can be a custom resource also. Something defined using the create-custom-resource command. Refer to one of the devtests I have written for an example.

mskdeepak-oracle commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+    }
+
+    /**
+     * If the message destination ref is linked to a message destination, fetch the linked destination and validate it.
```
I'm not sure what you are talking about. Didn't we realise that the problem is in tests?

mskdeepak-oracle commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,696 @@
+     * @return the converted name with java:global JNDI prefix.
+     */
+    private String convertModuleOrAppJNDIName(String jndiName, JndiNameEnvironment env) {
+        BundleDescriptor bd = null;
```
It is possible. But I have written code for the only two cases we are concerned about.

mskdeepak-oracle commented on appserver/resources/resources-runtime/src/main/java/org/glassfish/resources/module/ResourcesDeployer.java:
```diff
@@ -278,16 +279,32 @@ private void processArchive(DeploymentContext dc) {
                             ResourcesXMLParser.getNonConnectorResourcesList(list, false, true);
                     resourcesList.put(NON_CONNECTOR_RESOURCES, nonConnectorResources);
-                    
+                    // TODO: Some resources store their JNDI names in other attributes
```
Yes.

mskdeepak-oracle committed:
Comment out MDB

mskdeepak-oracle committed:
EJB Local refs validation done

mskdeepak-oracle committed:
Add testcase for EJB local refs

mskdeepak-oracle committed:
WebServiceRef Done. Need to test

mskdeepak-oracle committed:
Fix TODO's

mskdeepak-oracle committed:
Remote refs started. WIP

mskdeepak-oracle committed:
Fix devtests failing because of EJB validation

mskdeepak-oracle committed:
Add switch - deployment.switchoff.resource.validation to skip resource validation

mskdeepak-oracle committed:
Copy one more case from EjbNamingRefMan, include TODO's

mskdeepak-oracle committed:
Add a devtest for remote refs, fix bug in EJB resolving code

honghzzhang commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,681 @@
+            application = dc.getModuleMetaData(Application.class);
+            DeployCommandParameters commandParams = dc.getCommandParameters(DeployCommandParameters.class);
+            target = commandParams.target;
+            if (application == null || Boolean.getBoolean("deployment.switchoff.resource.validation"))
```
Is this a temporary system property till the feature implementation/testing is complete?

mskdeepak-oracle replied:
Yes. We want to use this in the CTS tests.

honghzzhang commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,681 @@
+
+        Server svr = domain.getServerNamed(target);
+        if (svr != null) {
+            return svr.isResourceRefExists(jndiName);
```
Maybe we should also check whether the enabled attribute is true here.

mskdeepak-oracle replied:
We have discussed this and came to a conclusion that the resources being configured is sufficient.

honghzzhang commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,681 @@
+        }
+
+        Cluster cluster = domain.getClusterNamed(target);
+        return cluster != null && cluster.isResourceRefExists(jndiName);
```
Same question here for the enabled attribute. If a resource is deployed but disabled on certain target, it will not be available.

honghzzhang commented on appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java:
```diff
@@ -0,0 +1,681 @@
+ * Created by Krishna Deepak on 6/9/17.
+ */
+@Service
+public class ResourceValidator implements EventListener, ResourceValidatorVisitor {
```
A general comment. In the past, this type of the resource validation happens at runtime instead of deployment time. You have probably already done this, but check the naming manager code which does the runtime check to see if we can re-use some of the logic instead of duplicating similar logic.

mskdeepak-oracle replied:
In general, I referred to the logic in `ComponentEnvManagerImpl` class while writing this code. Are you referring to the code in `GlassfishNamingMangerImpl`? I'm not sure how that can be reused, but I did copy the overall logic from a lot of different places into this class.

mskdeepak-oracle committed:
Minor changes

mskdeepak-oracle committed:
Fix bug. Major restructuring

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_all_resources

mskdeepak-oracle committed:
New test case for the bug

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_other_annotations

mskdeepak-oracle committed:
Merge with latest changes from validate_all_resources

mskdeepak-oracle committed:
Typo issue in error message text

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_all_resources

mskdeepak-oracle committed:
Get log message from local strings file, change sys prop to deployment.resource.validation

mskdeepak-oracle commented:
@glassfishrobot Run CI Tests please

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_other_annotations

mskdeepak-oracle committed:
Merge validate_other_annotations into main branch. All changes in same PR

mskdeepak-oracle committed:
Merge remote-tracking branch 'upstream/master' into validate_all_resources

mskdeepak-oracle committed:
Fix E1 class failures: Skip validation for non-managed bean classes in an EJB

mskdeepak-oracle committed:
Cleanup test folders

mskdeepak-oracle commented:
@glassfishrobot Run CI Tests please

mskdeepak-oracle commented:
@glassfishrobot Run CI Tests please
